### DISCUSSION TITLE
Question: How to produce single updates in an IObservableCache when batch editing a SourceCache

### DISCUSSION THREAD

**Discussion by Kl-Li** [reactivemarbles/DynamicData#570] [Q&A] [answered]
Hello, 
Is it possible to have only the latest update event (per key) instead of multiple in case of batch editing when using a `SourceCache` and an `IObservableCache`?

I use batch edit to add some market prices in the SourceCache (as shown below)
```
    public void AddTicks(ICollection<Tick> newTicks)
    {
        _ticksCache.Edit(innerCache => innerCache.AddOrUpdate(newTicks));
    }
```

I also have an `IObservableCache<Tick, string> NewTicks`  (as shown below)
```
    NewTicks = _ticksCache
        .Connect()
        .Filter(tick => tick.LastUpdate > DateTime.UtcNow - TimeSpan.FromSeconds(1))
        .Batch(TimeSpan.FromSeconds(1))
        .AsObservableCache();
```

When i subscribe to the `IObservableCache` i get multiple update events for the same currency pairs eg EURUSD per second
```
    tickService.NewTicks
        .Connect()
        .Watch("EURUSD")
    .Subscribe(price => Console.WriteLine(price));
```

In my case I'm only interested about the latest value of a currency pair (eg if EURUSD had 3 price changes in 1 second i'm only interested about the 3rd).
> Update, Key: EURUSD, Current: Name: EURUSD, Price: 1, LastUpdate: 14/03/2022 4:14:21 pm, Previous: Name: EURUSD, Price: 0, LastUpdate: 14/03/2022 4:14:20 pm
> Update, Key: EURUSD, Current: Name: EURUSD, Price: 2, LastUpdate: 14/03/2022 4:14:21 pm, Previous: Name: EURUSD, Price: 1, LastUpdate: 14/03/2022 4:14:21 pm
> Update, Key: EURUSD, Current: Name: EURUSD, Price: 3, LastUpdate: 14/03/2022 4:14:21 pm, Previous: Name: EURUSD, Price: 2, LastUpdate: 14/03/2022 4:14:21 pm

Is what I'm asking possible? If yes how?

ps: Yesterday I asked this question in [stackoverflow](https://stackoverflow.com/questions/71471772/how-to-produce-single-updates-when-batch-editing-in-dynamic-data). There i have my full sample code. I can copy it here as well if it's preferable. Also assuming i get an answer i will update/close this question accordingly (and vice versa of course).

**Comment by RolandPheasant** [comment]
Actually this is a long standing request which has never materialized into an operator.  There is some partial code already in the solution which aims to do just this but has never bubbled to the surface.

I have added the operator in this commit [EnsureUniqueKeys operator](https://github.com/reactivemarbles/DynamicData/commit/8da3f005bb57f215347dd73cf2bef9f27a7887e3#diff-eceb6212a4ecdb2e9e5efab619dbe8d3b8d285167fff1bfdd954c479c3254660) and will merge it into the main solution as soon as I have created some unit tests.

**Comment by RolandPheasant** [comment, marked as answer]
PR here https://github.com/reactivemarbles/DynamicData/pull/591

**Reply by RolandPheasant** [reply]
As per the example above, this will achieve unique changes

```cs
 NewTicks = _ticksCache
        .Connect()
        .Filter(tick => tick.LastUpdate > DateTime.UtcNow - TimeSpan.FromSeconds(1))
        .Batch(TimeSpan.FromSeconds(1))
        .EnsureUniqueKeys()       //NEW OPERATOR
        .AsObservableCache();
```

**Reply by Kl-Li** [reply]
Thank you very much @RolandPheasant 
I leave a link to your stack overflow answer here in case someone finds it usefull in the future. 
https://stackoverflow.com/a/72346359/13535877
